### PR TITLE
#1286: add configuration option to allow linking to section titles without explicit labels

### PR DESCRIPTION
**Pull request by tadhg-ohiggins**
There are a number of reasons why it should be possible to use :ref: to target section titles without requiring explicit labels:
-   The labels are often redundant (i.e. the same text as the titles).
-   If they are redundant, they clutter up the source significantly.
-   They make linking between documents more difficult, as the author must remember both to add the label to the target and, later, what the label was.
-   By making cross-referencing more difficult, they make it harder for authors to create navigable documentation.

A number of these were mentioned in issue #1286 and I hope this PR will reverse the WONTFIX decision.

My own use case is a cast of characters, each of which has a section. Adding an explicit label for each one is entirely redundant, and the argument that mandating explicit labels protect against future section title changes simply doesn't apply.

I'm sure many similar use cases exist, and that these are sufficient to warrant the addition of a configuration option to allow the non-explicit behavior, particularly since, as far as I can tell, the code change is quite simple.

This PR introduces a new configuration option named section_titles_as_targets and sets its default as False, preserving the existing behavior. If that option is set to True, the new behavior applies and the if statement that excludes non-explicit links in domains/std.py is effectively ignored.

### CODE REVIEW

File: sphinx/domains/std.py [hunk 1]
```diff
@@ -530,7 +530,8 @@ def process_doc(self, env, docname, document):
         labels, anonlabels = self.data['labels'], self.data['anonlabels']
         for name, explicit in iteritems(document.nametypes):
             if not explicit:
-                continue
+                if not env.config.section_titles_as_targets:
```
Comment (lehmannro) [inline sphinx/domains/std.py]: Combine into `if not explicit and not env.…`.

File: doc/config.rst [hunk 2]
```diff
@@ -171,6 +171,21 @@ General configuration
 
    .. versionadded:: 1.0
 
+.. confval:: section_titles_as_targets
+
+   The default is ``False``.
```
Comment (lehmannro) [inline doc/config.rst]: Starting with the default value is very odd.  I'd rephrase the whole section here… first, say what this setting is for, then what the values will do, and close with the default value.

File: doc/config.rst [hunk 3]
```diff
@@ -171,6 +171,21 @@ General configuration
...
+
+   If ``False``, explicit labels are necessary to use ``:ref:`` to link to section
```
Comment (lehmannro) [inline doc/config.rst:179]: `:rst:role:`ref`` instead.  (Also below.  Could drop the `See :ref:`ref-role`.`)

File: doc/markup/inline.rst [hunk 4]
```diff
@@ -154,6 +154,21 @@ Cross-referencing arbitrary locations
    section headings are changed, and for all builders that support
    cross-references.
 
+   .. versionadded:: 1.4
```
Comment (lehmannro) [inline doc/markup/inline.rst:157]: I'd have to try myself, but I've got a hunch that you should indent the following paragraphs as to belong to the `.. versionadded` directive.

File: sphinx/config.py [hunk 5]
```diff
@@ -74,6 +74,7 @@ class Config(object):
         modindex_common_prefix = ([], 'html'),
         rst_epilog = (None, 'env'),
         rst_prolog = (None, 'env'),
+        section_titles_as_targets = (None, 'env'),
```
Comment (lehmannro) [inline sphinx/config.py]: Shouldn't this be `False`?

File: doc/markup/inline.rst (hunk 4 above)
Comment (tadhg-ohiggins) [inline doc/markup/inline.rst:157]: Judging by the other examples in the doc, and having looked at the output, this doesn't appear to be the case.
I've incorporated your other suggestions; thanks for the feedback!

**Comment by tk0miya** [comment]
I investigated about what is `unexpected behavior is likely to occur`.
- If same section-name are used in same file:
  - no warnings are shown
  - could not refer them
- If same section-name are used in multiple files
  - warnings are shown: duplicate label test, other instance in .../foo.rst
  - can refer the last document (maybe sorted filename by alphabetically)

**Comment by tk0miya** [comment]
The following handler for `doctree-read` event regulates these warnings and referring behaviors:

``` python
def on_doctree_read(app, document):
    if not section_titles_as_targets:
        return

    labels = app.env.domaindata['std']['labels']
    anonlabels = app.env.domaindata['std']['anonlabels']
    for node in document.traverse(nodes.section):
        name = normalize_name(node[0].astext())
        labelid = node['ids'][0]
        docname = app.env.docname
        sectname = clean_astext(node[0])

        if name in labels:
            app.env.warn_node('duplicate label %s, ' % name + 'other instance '
                              'in ' + app.env.doc2path(labels[name][0]), node)

        anonlabels[name] = docname, labelid
        labels[name] = docname, labelid, sectname
```

**Comment by shimizukawa** [comment]
I think it would be better if the `on_doctree_read` implementation is provided as a bundled sphinx extension `sphinx.ext.section_titles_as_targets` as experimental feature.

**Comment by tk0miya** [comment]
+1 for bundled extension.
I'd prefer `sphinx.ext.autosectionlabel` as its new name.

@tadhg-ohiggins What do you think our proposal?
If we agree with this, I will arrange this to new extension.

**Comment by tadhg-ohiggins** [comment]
If a bundled extension makes more sense than a configuration option, that's fine with me. `sphinx.ext.autosectionlabel` works as a name, too.

If you could move the code to a bundled extension, that would be great.

Once it's in place I'm happy to clean up the documentation to fit into the extension documentation format, and to add the details you added about what goes wrong when there are duplicate section names.

**Comment by tk0miya** [comment]
Now, I add `sphinx.ext.autosectionlabel` into master branch.
Sphinx-1.4a1 will be released soon. Please try it :-)

@tadhg-ohiggins Your pull request has become of the wake of the feature.
Thank you for great proposal!
